File: androidApp/src/main/java/com/badmintontracker/android/localanalysis/SkeletonPanel.kt
package com.badmintontracker.android.localanalysis

import android.view.LayoutInflater
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.SeekParameters
import androidx.media3.ui.PlayerView
import com.badmintontracker.analysis.geometry.Matrix3x3
import com.badmintontracker.analysis.geometry.homography
import com.badmintontracker.analysis.geometry.maxResidualM
import com.badmintontracker.analysis.player.MetricKind
import com.badmintontracker.analysis.player.NearPlayerSelector
import com.badmintontracker.analysis.player.nearestPose
import com.badmintontracker.analysis.player.poseMetrics
import com.badmintontracker.analysis.player.poseToleranceS
import com.badmintontracker.android.R
import com.badmintontracker.android.clipdetail.FrameStepBar
import com.badmintontracker.android.clipdetail.PlaybackControlBar
import com.badmintontracker.shared.prefs.PlaybackPreferenceRepository
import com.badmintontracker.shared.prefs.RacketArmPreferenceRepository
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * What the file's court marks came to.
 *
 * [NONE] and [BAD] both leave the two court-plane tiles off the strip, but they
 * are different problems with different fixes - a re-run for marks that were
 * never stored, a re-marking for marks that do not fit - so the summary line
 * tells them apart rather than sending a coach to re-run a court it will
 * reject again.
 */
private enum class CourtFit { NONE, BAD, OK }

/** Whether the stored skeleton has been read from disk yet. */
private sealed interface SkeletonLoad {
    data object Loading : SkeletonLoad
    data class Loaded(
        val stored: SkeletonStore.Stored?,
        /** The resolved court fit from the file's marks; non-null only when [courtFit] is OK. */
        val homography: Matrix3x3?,
        val courtFit: CourtFit,
        /** One entry per pose, in pose order, for the graph. */
        val series: List<MetricSample>,
    ) : SkeletonLoad
}

/**
 * The near player's skeleton over the video it was measured on.
 *
 * Two things have to exist: the poses a run kept (only when the skeleton
 * metric was ticked) and the file that run decoded. Either missing is said in
 * one line rather than drawn around, because a skeleton over the wrong frames
 * looks like tracking that is broken, and a coach cannot tell that from
 * tracking that is bad.
 *
 * Under the video sit the strip and the graph: one tile per measurement of
 * the frame on screen, and the selected tile's kind drawn on the skeleton as
 * an arc and plotted over the two seconds either side of the playhead.
 * Nothing here is stored - every number is computed from the poses in the file
 * at view time, so a metric added later needs no re-run, and the two
 * court-plane tiles appear only when the file carries marks that fit.
 */
@Composable
fun SkeletonPanel(
    entryId: String,
    videoUri: String?,
    runner: LocalAnalysisRunner,
    prefs: PlaybackPreferenceRepository,
    racketArmPrefs: RacketArmPreferenceRepository,
    modifier: Modifier = Modifier,
) {
    // SkeletonStore.load parses the whole file, and its own KDoc prices a
    // 30-minute match at 11MB of poses - reading and decoding that on the
    // composition thread would jank the screen open. Loaded off the main
    // thread instead, with a one-line placeholder while it runs.
    // The metrics for every pose are computed here too, off the main thread
    // and once per file: the graph needs the whole series, and a court fit
    // whose marks miss by more than the selector's own gate is dropped rather
    // than turned into a stance in metres that is metres wrong.
    val load by produceState<SkeletonLoad>(initialValue = SkeletonLoad.Loading, key1 = entryId) {
        value = withContext(Dispatchers.IO) {
            val stored = runner.storedSkeleton(entryId)
            val marks = stored?.marks
            val homography = marks?.let { m ->
                m.homography()?.takeIf { h ->
                    val residual = m.maxResidualM(h)
                    residual != null && residual <= NearPlayerSelector.MAX_COURT_RESIDUAL_M
                }
            }
            val courtFit = when {
                marks == null -> CourtFit.NONE
                homography == null -> CourtFit.BAD
                else -> CourtFit.OK
            }
            val series = stored?.poses?.map {
                MetricSample(it.timestamp, poseMetrics(it.keypoints, it.confidence, homography))
            } ?: emptyList()
            SkeletonLoad.Loaded(stored, homography, courtFit, series)
        }
    }
    val source = remember(entryId, videoUri) { videoUri?.let { runner.analysedSource(entryId, it) } }

    Column(modifier = modifier.fillMaxSize().verticalScroll(rememberScrollState())) {
        when (val l = load) {
            SkeletonLoad.Loading -> Text("Loading skeleton", modifier = Modifier.padding(16.dp))
            is SkeletonLoad.Loaded -> when {
                l.stored == null -> Text(
                    "No skeleton was kept for this video. Run the analysis again with " +
                        "\"Skeleton playback\" ticked.",
                    modifier = Modifier.padding(16.dp),
                )
                source == null -> Text(
                    "The video this skeleton was measured on is no longer on this phone.",
                    modifier = Modifier.padding(16.dp),
                )
                else -> SkeletonPlayer(
                    source = source,
                    stored = l.stored,
                    homography = l.homography,
                    courtFit = l.courtFit,
                    series = l.series,
                    prefs = prefs,
                    racketArmPrefs = racketArmPrefs,
                    entryId = entryId,
                )
            }
        }
    }
}

/**
 * The player, the overlay, and the loop that keeps them on the same frame.
 *
 * The box takes the video's own aspect ratio, as the court-marking screen
 * does, so the player fills it edge to edge and the overlay's letterbox
 * arithmetic reduces to the identity. The position is read once per
 * display frame with `withFrameNanos`, because Media3 has no per-frame
 * position callback and a coach stepping frame by frame needs the joints to
 * move with the frame, not a few hundred milliseconds after it.
 */
@Composable
private fun SkeletonPlayer(
    source: File,
    stored: SkeletonStore.Stored,
    homography: Matrix3x3?,
    courtFit: CourtFit,
    series: List<MetricSample>,
    prefs: PlaybackPreferenceRepository,
    racketArmPrefs: RacketArmPreferenceRepository,
    entryId: String,
) {
    if (stored.videoWidth <= 0 || stored.videoHeight <= 0) {
        // Guards aspectRatio() below, which throws on a non-positive ratio.
        // A stored size this broken is another "this skeleton cannot be
        // shown" case, not a crash.
        Text(
            "This skeleton's stored video size is invalid. Run the analysis again to rebuild it.",
            modifier = Modifier.padding(16.dp),
        )
        return
    }

    val context = LocalContext.current
    val player = remember {
        ExoPlayer.Builder(context).build().apply { setSeekParameters(SeekParameters.EXACT) }
    }
    var playbackError by remember { mutableStateOf<String?>(null) }
    DisposableEffect(player) {
        // Mirrors LocalPlayerScreen: the analysed file can be moved or
        // deleted behind us, and without a listener the player just
        // freezes silently instead of saying so.
        val listener = object : Player.Listener {
            override fun onPlayerError(error: PlaybackException) {
                playbackError = "Couldn't play the analysed video. Run the analysis again to rebuild it."
            }
        }
        player.addListener(listener)
        onDispose {
            player.removeListener(listener)
            player.release()
        }
    }
    LaunchedEffect(source) {
        player.setMediaItem(MediaItem.fromUri(android.net.Uri.fromFile(source)))
        player.prepare()
        player.playWhenReady = false
    }

    var positionMs by remember { mutableLongStateOf(0L) }
    LaunchedEffect(player) {
        while (true) {
            // Polls once per display frame rather than waiting on a Media3
            // callback, because there is none. This keeps the Recomposer
            // non-idle for as long as this screen is composed, so a Compose
            // UI test of this screen must not wait for idle: it never will.
            withFrameNanos { }
            positionMs = player.currentPosition
        }
    }
    val tolerance = remember(stored.fps) { poseToleranceS(stored.fps) }
    val pose = remember(positionMs, stored) { nearestPose(stored.poses, positionMs / 1000.0, tolerance) }

    val hasCourt = homography != null
    var racketArm by remember(entryId) { mutableStateOf(racketArmPrefs.racketArm(entryId)) }
    var chosen by rememberSaveable(entryId) {
        mutableStateOf(if (hasCourt) MetricKind.STANCE else MetricKind.ELBOW_RIGHT)
    }
    val visible = visibleKinds(hasCourt, racketArm)
    // The coach's choice held to the tiles on screen: the racket-arm control can
    // hide the chosen kind, and the graph and the arc must never show a kind with
    // no tile. Derived, not written back, so composition stays a read; the choice
    // itself is kept, so putting the arm back brings the tile back selected.
    val selected = if (chosen in visible) chosen else visible.first()
    val metrics = remember(pose, homography) { pose?.let { poseMetrics(it.keypoints, it.confidence, homography) } }

    val error = playbackError
    if (error != null) {
        // In place of the video box: the overlay is inside it and so never
        // draws when this branch runs instead. The transport bars stay -
        // seeking or pausing an errored ExoPlayer is a no-op, not a crash -
        // but the summary line below is skipped explicitly, since it too is
        // about a frame that is not on screen.
        Text(error, modifier = Modifier.padding(16.dp))
    } else {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .aspectRatio(stored.videoWidth.toFloat() / stored.videoHeight.toFloat())
                .background(Color.Black),
        ) {
            AndroidView(
                factory = { c ->
                    // The layout, not PlayerView(c): it asks for a texture surface, and
                    // only a texture surface lets Compose animate and scroll over the
                    // video. A SurfaceView renders in its own window, so it survives the
                    // exit animation for a frame on top of the next screen (715e22b),
                    // and this one is inside a scroller as well.
                    val view = LayoutInflater.from(c)
                        .inflate(R.layout.clip_player_view, null) as PlayerView
                    view.apply {
                        this.player = player
                        // The bars below own transport; a controller over the
                        // skeleton would sit exactly where the joints are.
                        useController = false
                    }
                },
                modifier = Modifier.fillMaxSize(),
            )
            if (pose != null) {
                SkeletonOverlay(
                    keypoints = pose.keypoints,
                    confidence = pose.confidence,
                    videoWidth = stored.videoWidth,
                    videoHeight = stored.videoHeight,
                    modifier = Modifier.fillMaxSize(),
                    highlight = selected.highlight(),
                )
            }
        }
        MetricsStrip(
            metrics = metrics,
            hasCourt = hasCourt,
            racketArm = racketArm,
            onRacketArm = { arm ->
                racketArm = arm
                racketArmPrefs.setRacketArm(entryId, arm)
            },
            selected = selected,
            onSelect = { chosen = it },
        )
    }
    PlaybackControlBar(player = player, prefs = prefs)
    FrameStepBar(player = player)
    if (error == null) {
        MetricGraph(
            series = series,
            kind = selected,
            positionS = positionMs / 1000.0,
            fps = stored.fps,
            durationS = player.duration.takeIf { it > 0 }?.let { it / 1000.0 } ?: Double.POSITIVE_INFINITY,
            onSeek = { seconds ->
                if (player.isPlaying) player.pause()
                player.seekTo(
                    (seconds * 1000).toLong()
                        .coerceIn(0L, player.duration.takeIf { it > 0 } ?: Long.MAX_VALUE),
                )
            },
        )
        // The court line is part of the summary because a file whose marks are
        // missing or do not fit is missing two tiles, and a coach should not
        // have to guess why, nor which of the two fixes to reach for.
        val court = when (courtFit) {
            CourtFit.OK -> " · court marks in file"
            CourtFit.NONE -> " · no court marks: stance and position need a re-run"
            CourtFit.BAD -> " · court marks do not fit: mark the court again for stance and position"
        }
        Text(
            "Skeleton in ${stored.poses.size} frames" +
                (pose?.let { " · frame ${it.frame}" } ?: " · no skeleton at this frame") + court,
            style = MaterialTheme.typography.labelSmall,
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
        )
    }
}

File: iosApp/Sources/LocalAnalysis/SkeletonPanel.swift
import AVFoundation
import Shared
import SwiftUI

/// Everything the panel reads that does not change with the playhead, plus the
/// playhead itself.
///
/// One observable object rather than a screenful of `@State`, because the
/// position moves thirty times a second: held in the view, every tick would
/// re-evaluate the body that owns the loaded skeleton and the series, and
/// neither is cheap to hand back to SwiftUI. Held here, a tick invalidates only
/// the views that actually read `position`.
///
/// The series stays the Kotlin list `poseMetrics` produced. Mapping it into
/// Swift values would be an O(poses) bridge crossing, and `graphSegments` - the
/// only thing that reads it - takes the Kotlin list anyway.
@Observable @MainActor
final class SkeletonPanelModel {
    enum Load {
        case loading
        case loaded(Loaded)
    }

    struct Loaded {
        let stored: SkeletonStore.Stored?
        /// The resolved court fit from the file's marks; non-nil only when
        /// `courtFit` is `.ok`.
        let homography: [[KotlinDouble]]?
        let courtFit: CourtFit
        /// One entry per pose, in pose order, for the graph.
        let series: [MetricSample]
    }

    private(set) var load: Load = .loading
    private(set) var position = PlaybackPosition()
    /// The pose the playhead is on, and what it measures.
    ///
    /// Resolved when the position moves rather than read from the view, and only
    /// when the FRAME changes: the playhead ticks thirty times a second, a frame
    /// lasts one thirtieth, and `poseMetrics` is a full pass over seventeen
    /// joints while `nearestPose` hands the whole pose list back across the
    /// Objective-C bridge. androidApp gets the same for free from
    /// `remember(positionMs, stored)` and `remember(pose, homography)`.
    private(set) var pose: PlayerPose?
    private(set) var metrics: PoseMetrics?
    private var resolvedFrame: Int32? = nil
    var racketArm: RacketArm?
    /// The coach's pick, kept even while the racket-arm control hides its tile -
    /// see `selected`.
    var chosen: MetricKind = .elbowRight
    var expanded = false
    var speed = PlaybackOptions.shared.DEFAULT_SPEED
    var playbackError: String? = nil
    private(set) var player: AVPlayer?
    /// The video's own frame rate, for whole-frame stepping. Read from the
    /// asset rather than from the skeleton file: the stored fps is the analysis
    /// pass's, which is the same number, but the transport is stepping THIS
    /// player's media.
    private var fps: Float = 0
    private var failureObservation: NSKeyValueObservation?

    let entryId: String
    private let runner: LocalAnalysisRunner
    private let racketArmPrefs: RacketArmPreferenceRepository

    init(entryId: String, runner: LocalAnalysisRunner, racketArmPrefs: RacketArmPreferenceRepository) {
        self.entryId = entryId
        self.runner = runner
        self.racketArmPrefs = racketArmPrefs
        self.racketArm = racketArmPrefs.racketArm(entryId: entryId)
    }

    /// Reads the file and measures every pose in it, off the main actor.
    ///
    /// `SkeletonStore.load` parses the whole file, and its own note prices a
    /// 30-minute match at 11MB of poses; measuring them is another pass over the
    /// same 6,000 frames. Doing either while the screen opens would jank it, so
    /// both happen here with a one-line placeholder up in the meantime.
    ///
    /// A court fit whose marks miss by more than the near-player selector's own
    /// gate is dropped rather than turned into a stance in metres that is metres
    /// wrong.
    func loadSkeleton(videoRelativePath: String?) async {
        let entryId = self.entryId
        let runner = self.runner
        let loaded = await Task.detached(priority: .userInitiated) { () -> Loaded in
            let stored = runner.storedSkeleton(entryId: entryId)
            let marks = stored?.marks
            let homography = marks?.homography().flatMap { candidate -> [[KotlinDouble]]? in
                guard let residual = marks?.maxResidualM(h: candidate)?.doubleValue,
                      residual <= NearPlayerSelector.companion.MAX_COURT_RESIDUAL_M
                else { return nil }
                return candidate
            }
            let courtFit: CourtFit = marks == nil ? .none : (homography == nil ? .bad : .ok)
            let series = (stored?.poses ?? []).map { pose in
                MetricSample(
                    timestamp: pose.timestamp,
                    metrics: PoseMetricsKt.poseMetrics(
                        keypoints: pose.keypoints,
                        confidence: pose.confidence,
                        homography: homography,
                        minConfidence: NearPlayerSelector.companion.MIN_KEYPOINT_CONFIDENCE
                    )
                )
            }
            return Loaded(stored: stored, homography: homography, courtFit: courtFit, series: series)
        }.value

        load = .loaded(loaded)
        chosen = loaded.courtFit == .ok ? .stance : .elbowRight
        // The first tick is a thirtieth of a second away and the panel draws
        // before it: without this the video's opening frame has no skeleton on
        // it and every tile reads as a dash.
        resolvedFrame = nil
        resolvePose()
        guard loaded.stored != nil,
              let videoRelativePath,
              let url = runner.analysedSource(relativePath: videoRelativePath)
        else { return }
        preparePlayer(url: url)
        await loadFrameRate(url: url)
    }

    private func preparePlayer(url: URL) {
        let created = AVPlayer(url: url)
        // Mirrors LocalPlayerModel: the analysed file can be moved or deleted
        // behind us, and without this the player just freezes silently instead
        // of saying so.
        failureObservation = created.observe(\.currentItem?.status) { [weak self] player, _ in
            guard player.currentItem?.status == .failed else { return }
            Task { @MainActor [weak self] in
                self?.playbackError =
                    "Couldn't play the analyzed video. Run the analysis again to rebuild it."
            }
        }
        player = created
    }

    private func loadFrameRate(url: URL) async {
        let asset = AVURLAsset(url: url)
        if let track = try? await asset.loadTracks(withMediaType: .video).first,
           let rate = try? await track.load(.nominalFrameRate) {
            fps = rate
        }
    }

    /// Moves the playhead, and with it the pose and the measurements - but only
    /// when the frame under it actually changed. See `pose`.
    func setPosition(_ next: PlaybackPosition) {
        position = next
        resolvePose()
    }

    private func resolvePose() {
        guard case .loaded(let loaded) = load, let stored = loaded.stored else { return }
        let found = PoseLookupKt.nearestPose(
            poses: stored.poses,
            seconds: Double(position.positionMs) / 1000,
            toleranceS: PoseLookupKt.poseToleranceS(fps: stored.fps)
        )
        // Nil is a frame with no pose and is a state of its own, so "no change"
        // has to compare the two absences too, not just two frame numbers.
        guard found?.frame != resolvedFrame || (found == nil) != (pose == nil) else { return }
        resolvedFrame = found?.frame
        pose = found
        metrics = found.map { found in
            PoseMetricsKt.poseMetrics(
                keypoints: found.keypoints,
                confidence: found.confidence,
                homography: loaded.homography,
                minConfidence: NearPlayerSelector.companion.MIN_KEYPOINT_CONFIDENCE
            )
        }
    }

    var hasCourt: Bool {
        guard case .loaded(let loaded) = load else { return false }
        return loaded.homography != nil
    }

    /// The coach's choice held to the tiles on screen: the racket-arm control
    /// can hide the chosen kind, and the graph and the arc must never show a
    /// kind with no tile. Derived rather than written back, so the choice itself
    /// is kept and putting the arm back brings the tile back selected.
    var selected: MetricKind {
        let visible = MetricsFormatKt.visibleKinds(hasCourt: hasCourt, racketArm: racketArm)
        return visible.contains(chosen) ? chosen : (visible.first ?? .elbowRight)
    }

    func setRacketArm(_ arm: RacketArm?) {
        racketArm = arm
        racketArmPrefs.setRacketArm(entryId: entryId, arm: arm)
    }

    func stepFrames(_ delta: Int64) {
        guard let player else { return }
        let duration = player.currentItem?.duration
        player.seek(
            to: CMTime(
                seconds: FrameStepMath.targetSeconds(
                    currentSeconds: player.currentTime().isNumeric
                        ? CMTimeGetSeconds(player.currentTime()) : 0,
                    fps: fps,
                    delta: delta,
                    durationSeconds: (duration?.isNumeric == true) ? CMTimeGetSeconds(duration!) : 0
                ),
                preferredTimescale: 600
            ),
            toleranceBefore: .zero,
            toleranceAfter: .zero
        )
    }

    /// Seeks from the graph, which asks in absolute seconds rather than in a
    /// fraction. Pauses first: scrubbing a curve while the video runs away under
    /// the finger is not scrubbing.
    func seek(toSeconds seconds: Double) {
        guard let player else { return }
        if player.rate != 0 { player.pause() }
        let duration = player.currentItem?.duration
        let limit = (duration?.isNumeric == true) ? CMTimeGetSeconds(duration!) : .greatestFiniteMagnitude
        player.seek(
            to: CMTime(seconds: min(max(seconds, 0), limit), preferredTimescale: 600),
            toleranceBefore: .zero,
            toleranceAfter: .zero
        )
    }
}

/// The near player's skeleton over the video it was measured on.
///
/// Two things have to exist: the poses a run kept (only when the skeleton metric
/// was ticked) and the file that run decoded. Either missing is said in one line
/// rather than drawn around, because a skeleton over the wrong frames looks like
/// tracking that is broken, and a coach cannot tell that from tracking that is
/// bad.
///
/// Under the video sit the transport, the strip and the graph: one tile per
/// measurement of the frame on screen, and the selected tile's kind drawn on the
/// skeleton as an arc and plotted over the two seconds either side of the
/// playhead. Nothing here is stored - every number is computed from the poses in
/// the file at view time, so a metric added later needs no re-run, and the two
/// court-plane tiles appear only when the file carries marks that fit.
///
/// Port of androidApp's `SkeletonPanel`.
struct SkeletonPanel: View {
    let entryId: String
    /// The entry's own path, or nil when the video it was measured on is gone.
    let videoRelativePath: String?
    let runner: LocalAnalysisRunner
    let prefs: PlaybackPreferenceRepository
    let racketArmPrefs: RacketArmPreferenceRepository

    @State private var model: SkeletonPanelModel

    init(
        entryId: String,
        videoRelativePath: String?,
        runner: LocalAnalysisRunner,
        prefs: PlaybackPreferenceRepository,
        racketArmPrefs: RacketArmPreferenceRepository
    ) {
        self.entryId = entryId
        self.videoRelativePath = videoRelativePath
        self.runner = runner
        self.prefs = prefs
        self.racketArmPrefs = racketArmPrefs
        _model = State(initialValue: SkeletonPanelModel(
            entryId: entryId, runner: runner, racketArmPrefs: racketArmPrefs
        ))
    }

    var body: some View {
        content
            .task { await model.loadSkeleton(videoRelativePath: videoRelativePath) }
            .task {
                for await value in prefs.metricsExpanded {
                    model.expanded = value.boolValue
                }
            }
    }

    @ViewBuilder
    private var content: some View {
        switch model.load {
        case .loading:
            PanelMessage(text: "Loading skeleton")
        case .loaded(let loaded):
            if let stored = loaded.stored {
                if stored.videoWidth <= 0 || stored.videoHeight <= 0 {
                    // Guards the aspect ratio below, which a non-positive size
                    // would make meaningless. A stored size this broken is
                    // another "this skeleton cannot be shown" case.
                    PanelMessage(
                        text: "This skeleton's stored video size is invalid. "
                            + "Run the analysis again to rebuild it."
                    )
                } else if model.player == nil {
                    PanelMessage(
                        text: "The video this skeleton was measured on is no longer on this phone."
                    )
                } else {
                    player(stored: stored, loaded: loaded)
                }
            } else {
                PanelMessage(
                    text: "No skeleton was kept for this video. "
                        + "Run the analysis again with \"Skeleton playback\" ticked."
                )
            }
        }
    }

    private func player(stored: SkeletonStore.Stored, loaded: SkeletonPanelModel.Loaded) -> some View {
        let pose = model.pose
        return VStack(alignment: .leading, spacing: 0) {
            ShuttlPlayer(
                player: model.player,
                prefs: prefs,
                step: { model.stepFrames($0) },
                // The video's own stored ratio, so the player fills the card
                // edge to edge and the overlay's letterbox arithmetic reduces to
                // the identity.
                aspectRatio: Double(stored.videoWidth) / Double(stored.videoHeight),
                onPosition: { model.setPosition($0) },
                overlay: {
                    if let pose {
                        SkeletonOverlay(
                            keypoints: pose.keypoints,
                            confidence: pose.confidence.map(\.floatValue),
                            videoWidth: Int(stored.videoWidth),
                            videoHeight: Int(stored.videoHeight),
                            highlight: poseHighlight(for: model.selected)
                        )
                    }
                },
                errorContent: {
                    if let error = model.playbackError {
                        ErrorBanner(message: error).padding(16)
                    }
                }
            )
            // The frame is the one fact read per frame; the speed sits opposite
            // it, as the mock lays the line out.
            HStack(alignment: .firstTextBaseline) {
                Text(SkeletonSummaryKt.skeletonFooter(frame: pose.map { KotlinInt(int: $0.frame) }))
                    .shuttlType(ShuttlType.bodySmall)
                    .foregroundStyle(Shuttl.textTertiary)
                Spacer(minLength: 8)
                Text("\(PlaybackOptions.shared.formatSpeed(speed: model.speed)) speed")
                    .shuttlType(ShuttlType.bodySmall)
                    .foregroundStyle(Shuttl.textSecondary)
            }
            .padding(.horizontal, ShuttlGutter.page)
            .padding(.top, 22)
            MetricsStrip(
                metrics: model.metrics,
                hasCourt: model.hasCourt,
                racketArm: model.racketArm,
                onRacketArm: { model.setRacketArm($0) },
                selected: model.selected,
                onSelect: { model.chosen = $0 },
                expanded: model.expanded,
                onExpanded: { prefs.setMetricsExpanded(expanded: $0) },
                detail: SkeletonSummaryKt.skeletonDetail(
                    frames: Int32(stored.poses.count), courtFit: loaded.courtFit
                )
            )
            .padding(.top, 10)
            MetricGraph(
                series: loaded.series,
                kind: model.selected,
                label: MetricsFormatKt.metricLabel(kind: model.selected, racketArm: model.racketArm),
                positionS: Double(model.position.positionMs) / 1000,
                fps: stored.fps,
                durationS: model.position.durationMs > 0
                    ? Double(model.position.durationMs) / 1000
                    : .infinity,
                onSeek: { model.seek(toSeconds: $0) }
            )
            .padding(.top, 8)
            // A court warning is shown here, not only in the expanded detail: a
            // coach looking for the Stance tile should not have to open the grid
            // to learn why it is not there.
            if let warning = SkeletonSummaryKt.courtWarning(courtFit: loaded.courtFit) {
                Text(warning)
                    .shuttlType(ShuttlType.bodySmall)
                    .foregroundStyle(Shuttl.textTertiary)
                    .padding(.horizontal, ShuttlGutter.page)
                    .padding(.vertical, 12)
            }
            Spacer(minLength: ShuttlGutter.page)
        }
        .task {
            for await value in prefs.speed {
                model.speed = value.floatValue
            }
        }
    }
}
